File: src/core/process-registry.ts
/**
 * Process Registry
 *
 * Tracks spawned processes by session and provides cross-platform termination.
 * Enables reliable stop/pause/kill operations across CLI and GUI.
 *
 * See BUILD_QUEUE_IMPROVEMENTS.md P0-T21.
 */

import { promises as fs } from 'fs';
import { spawn } from 'child_process';
import { dirname } from 'path';

/**
 * Record of a single spawned process
 */
export interface ProcessRecord {
  /** Process ID */
  pid: number;
  /** Platform where process was spawned (cursor, claude, codex, etc.) */
  platform: string;
  /** Command that was executed */
  command: string;
  /** ISO timestamp when process was spawned */
  spawnedAt: string;
  /** Current status of the process */
  status: 'running' | 'terminated' | 'killed';
}

/**
 * Session registry containing all processes for a session
 */
export interface SessionRegistry {
  /** Unique session identifier */
  sessionId: string;
  /** PID of the orchestrator process */
  orchestratorPid: number;
  /** All processes spawned during this session */
  processes: ProcessRecord[];
  /** ISO timestamp when session started */
  startedAt: string;
  /** Current status of the session */
  status: 'active' | 'stopped' | 'crashed';
}

/**
 * ProcessRegistry class
 *
 * Manages process tracking and termination for a session.
 * Persists to `.puppet-master/sessions/{sessionId}.json`
 */
export class ProcessRegistry {
  private registry: SessionRegistry | null = null;

  /**
   * Creates a new ProcessRegistry instance.
   * @param sessionId - Session identifier
   * @param registryPath - Path to session registry file
   */
  constructor(
    private readonly sessionId: string,
    private readonly registryPath: string
  ) {}

  /**
   * Initializes the registry (loads existing or creates new).
   */
  async initialize(): Promise<void> {
    this.registry = await this.load();
    if (!this.registry) {
      this.registry = {
        sessionId: this.sessionId,
        orchestratorPid: process.pid,
        processes: [],
        startedAt: new Date().toISOString(),
        status: 'active',
      };
      await this.save();
    }
  }

  /**
   * Registers a new process.
   * @param pid - Process ID
   * @param platform - Platform name
   * @param command - Command that was executed
   */
  async registerProcess(pid: number, platform: string, command: string): Promise<void> {
    if (!this.registry) {
      await this.initialize();
    }

    const record: ProcessRecord = {
      pid,
      platform,
      command,
      spawnedAt: new Date().toISOString(),
      status: 'running',
    };

    this.registry!.processes.push(record);
    await this.save();
  }

  /**
   * Terminates a specific process.
   * @param pid - Process ID to terminate
   * @param force - Whether to force kill (SIGKILL on Unix, /F on Windows)
   */
  async terminateProcess(pid: number, force: boolean = false): Promise<void> {
    if (!this.registry) {
      await this.initialize();
    }

    // Find the process record
    const record = this.registry!.processes.find((p) => p.pid === pid);
    if (!record || record.status !== 'running') {
      return; // Already terminated or not found
    }

    try {
      await this.terminateProcessCrossPlatform(pid, force);
      record.status = force ? 'killed' : 'terminated';
      await this.save();
    } catch (error) {
      // Process may have already exited, mark as terminated
      record.status = 'terminated';
      await this.save();
    }
  }

  /**
   * Terminates all running processes in this session.
   * @param force - Whether to force kill all processes
   */
  async terminateAll(force: boolean = false): Promise<void> {
    if (!this.registry) {
      await this.initialize();
    }

    const runningProcesses = this.registry!.processes.filter((p) => p.status === 'running');

    // Terminate all running processes
    for (const record of runningProcesses) {
      await this.terminateProcess(record.pid, force);
    }

    // Update session status
    this.registry!.status = 'stopped';
    await this.save();
  }

  /**
   * Gets all running processes.
   * @returns Array of running process records
   */
  async getRunningProcesses(): Promise<ProcessRecord[]> {
    if (!this.registry) {
      await this.initialize();
    }

    return this.registry!.processes.filter((p) => p.status === 'running');
  }

  /**
   * Gets the full session registry.
   * @returns Session registry or null if not initialized
   */
  getRegistry(): SessionRegistry | null {
    return this.registry;
  }

  /**
   * Terminates a process using cross-platform method.
   * @param pid - Process ID to terminate
   * @param force - Whether to force kill
   */
  private async terminateProcessCrossPlatform(pid: number, force: boolean): Promise<void> {
    if (process.platform === 'win32') {
      // Windows: use taskkill
      return new Promise((resolve, reject) => {
        const signal = force ? '/F' : '/T';
        const proc = spawn('taskkill', [signal, '/PID', String(pid)], {
          stdio: 'ignore',
        });

        proc.on('close', (code) => {
          if (code === 0 || code === 128) {
            // 128 = process not found (already terminated)
            resolve();
          } else {
            reject(new Error(`taskkill exited with code ${code}`));
          }
        });

        proc.on('error', reject);
      });
    } else {
      // Unix: use process.kill with process groups
      const signal = force ? 'SIGKILL' : 'SIGTERM';
      try {
        // Try to kill process group (negative PID)
        process.kill(-pid, signal);
      } catch (err) {
        // Fallback: kill single process
        try {
          process.kill(pid, signal);
        } catch (innerErr) {
          // Process may have already exited
          const error = innerErr as NodeJS.ErrnoException;
          if (error.code !== 'ESRCH') {
            // ESRCH = No such process (already terminated)
            throw innerErr;
          }
        }
      }
    }
  }

  /**
   * Saves the registry to disk.
   */
  private async save(): Promise<void> {
    if (!this.registry) {
      return;
    }

    // Ensure directory exists
    await fs.mkdir(dirname(this.registryPath), { recursive: true });

    // Write registry file
    await fs.writeFile(this.registryPath, JSON.stringify(this.registry, null, 2), 'utf-8');
  }

  /**
   * Loads the registry from disk.
   * @returns Session registry or null if file doesn't exist
   */
  private async load(): Promise<SessionRegistry | null> {
    try {
      const content = await fs.readFile(this.registryPath, 'utf-8');
      const registry = JSON.parse(content) as SessionRegistry;

      // Validate structure
      if (
        !registry.sessionId ||
        !registry.processes ||
        !Array.isArray(registry.processes) ||
        !registry.startedAt
      ) {
        throw new Error('Invalid session registry structure');
      }

      return registry;
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
        return null; // File doesn't exist
      }
      throw error;
    }
  }
}

File: puppet-master-rs/src/core/process_registry.rs
//! Process registry for tracking and terminating spawned processes
//!
//! Tracks all child processes spawned during execution with:
//! - Cross-platform termination (SIGTERM/SIGKILL on Unix, TerminateProcess on Windows)
//! - Process metadata and status tracking
//! - Session-based registry persistence
//! - Clean shutdown support

use anyhow::{Context, Result};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;

#[cfg(unix)]
use nix::sys::signal::{kill, Signal};
#[cfg(unix)]
use nix::unistd::Pid;

/// Status of a tracked process
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ProcessStatus {
    /// Process is running
    Running,
    /// Process was gracefully terminated
    Terminated,
    /// Process was forcefully killed
    Killed,
    /// Process exited on its own
    Exited,
}

/// Record of a single spawned process
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProcessRecord {
    /// Process ID
    pub pid: u32,
    /// Platform that spawned this process
    pub platform: String,
    /// Command that was executed
    pub command: String,
    /// Arguments passed to command
    pub args: Vec<String>,
    /// When the process was spawned
    pub spawned_at: DateTime<Utc>,
    /// Current status
    pub status: ProcessStatus,
    /// When process ended (if not running)
    pub ended_at: Option<DateTime<Utc>>,
    /// Exit code (if exited)
    pub exit_code: Option<i32>,
}

/// Session-level process registry
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionRegistry {
    /// Session ID
    pub session_id: String,
    /// Orchestrator PID
    pub orchestrator_pid: u32,
    /// All tracked processes
    pub processes: Vec<ProcessRecord>,
    /// When session started
    pub started_at: DateTime<Utc>,
    /// Session status
    pub status: SessionStatus,
}

/// Status of a session
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SessionStatus {
    /// Session is active
    Active,
    /// Session was stopped gracefully
    Stopped,
    /// Session crashed or was interrupted
    Crashed,
}

/// Process registry manager
pub struct ProcessRegistry {
    /// Session ID
    session_id: String,
    /// Path to registry file
    registry_path: PathBuf,
    /// In-memory registry
    registry: SessionRegistry,
}

impl ProcessRegistry {
    /// Create new process registry
    ///
    /// # Arguments
    /// * `session_id` - Unique session identifier
    /// * `registry_path` - Path to persist registry (e.g., .puppet-master/sessions/{id}.json)
    pub fn new(session_id: String, registry_path: PathBuf) -> Self {
        let orchestrator_pid = std::process::id();
        
        let registry = SessionRegistry {
            session_id: session_id.clone(),
            orchestrator_pid,
            processes: Vec::new(),
            started_at: Utc::now(),
            status: SessionStatus::Active,
        };

        Self {
            session_id,
            registry_path,
            registry,
        }
    }

    /// Initialize registry (load existing or create new)
    pub async fn initialize(&mut self) -> Result<()> {
        // Try to load existing registry
        if self.registry_path.exists() {
            let json = fs::read_to_string(&self.registry_path)
                .context("Failed to read registry file")?;
            self.registry = serde_json::from_str(&json)
                .context("Failed to deserialize registry")?;
        }
        
        // Ensure directory exists
        if let Some(parent) = self.registry_path.parent() {
            fs::create_dir_all(parent)
                .context("Failed to create registry directory")?;
        }

        // Save initial state
        self.save().await?;

        Ok(())
    }

    /// Register a new process
    ///
    /// # Arguments
    /// * `pid` - Process ID
    /// * `platform` - Platform name (e.g., "cursor", "claude")
    /// * `command` - Command executed
    /// * `args` - Command arguments
    pub async fn register(
        &mut self,
        pid: u32,
        platform: impl Into<String>,
        command: impl Into<String>,
        args: Vec<String>,
    ) -> Result<()> {
        let record = ProcessRecord {
            pid,
            platform: platform.into(),
            command: command.into(),
            args,
            spawned_at: Utc::now(),
            status: ProcessStatus::Running,
            ended_at: None,
            exit_code: None,
        };

        self.registry.processes.push(record);
        self.save().await?;

        Ok(())
    }

    /// Unregister a process (mark as exited)
    ///
    /// # Arguments
    /// * `pid` - Process ID to unregister
    /// * `exit_code` - Exit code if known
    pub async fn unregister(&mut self, pid: u32, exit_code: Option<i32>) -> Result<()> {
        if let Some(record) = self.registry.processes.iter_mut().find(|p| p.pid == pid) {
            record.status = ProcessStatus::Exited;
            record.ended_at = Some(Utc::now());
            record.exit_code = exit_code;
            self.save().await?;
        }

        Ok(())
    }

    /// Terminate a specific process
    ///
    /// # Arguments
    /// * `pid` - Process ID to terminate
    /// * `force` - If true, use SIGKILL/TerminateProcess; if false, use SIGTERM
    pub async fn kill_process(&mut self, pid: u32, force: bool) -> Result<()> {
        // Find the process record
        let record = self.registry.processes.iter_mut().find(|p| p.pid == pid);
        
        if let Some(record) = record {
            if record.status != ProcessStatus::Running {
                // Already terminated
                return Ok(());
            }

            // Attempt termination
            let result = terminate_process(pid, force);
            
            match result {
                Ok(()) => {
                    record.status = if force {
                        ProcessStatus::Killed
                    } else {
                        ProcessStatus::Terminated
                    };
                    record.ended_at = Some(Utc::now());
                    self.save().await?;
                }
                Err(e) => {
                    // Process may have already exited
                    if is_process_not_found_error(&e) {
                        record.status = ProcessStatus::Exited;
                        record.ended_at = Some(Utc::now());
                        self.save().await?;
                    } else {
                        return Err(e);
                    }
                }
            }
        }

        Ok(())
    }

    /// Terminate all running processes
    ///
    /// # Arguments
    /// * `force` - If true, force kill all processes
    pub async fn kill_all(&mut self, force: bool) -> Result<()> {
        let running_pids: Vec<u32> = self
            .registry
            .processes
            .iter()
            .filter(|p| p.status == ProcessStatus::Running)
            .map(|p| p.pid)
            .collect();

        for pid in running_pids {
            // Ignore errors for individual processes
            let _ = self.kill_process(pid, force).await;
        }

        self.registry.status = SessionStatus::Stopped;
        self.save().await?;

        Ok(())
    }

    /// Kill processes by platform
    ///
    /// # Arguments
    /// * `platform` - Platform name to filter by
    /// * `force` - Whether to force kill
    pub async fn kill_by_platform(&mut self, platform: &str, force: bool) -> Result<()> {
        let matching_pids: Vec<u32> = self
            .registry
            .processes
            .iter()
            .filter(|p| p.platform == platform && p.status == ProcessStatus::Running)
            .map(|p| p.pid)
            .collect();

        for pid in matching_pids {
            let _ = self.kill_process(pid, force).await;
        }

        Ok(())
    }

    /// Get all currently running processes
    pub fn list_active(&self) -> Vec<&ProcessRecord> {
        self.registry
            .processes
            .iter()
            .filter(|p| p.status == ProcessStatus::Running)
            .collect()
    }

    /// Get all processes for a specific platform
    pub fn list_by_platform(&self, platform: &str) -> Vec<&ProcessRecord> {
        self.registry
            .processes
            .iter()
            .filter(|p| p.platform == platform)
            .collect()
    }

    /// Get full registry
    pub fn get_registry(&self) -> &SessionRegistry {
        &self.registry
    }

    /// Mark session as crashed
    pub async fn mark_crashed(&mut self) -> Result<()> {
        self.registry.status = SessionStatus::Crashed;
        self.save().await
    }

    /// Save registry to disk
    async fn save(&self) -> Result<()> {
        let json = serde_json::to_string_pretty(&self.registry)
            .context("Failed to serialize registry")?;
        
        // Atomic write using temp file
        let temp_path = self.registry_path.with_extension("tmp");
        fs::write(&temp_path, json)
            .context("Failed to write temp registry file")?;
        
        fs::rename(&temp_path, &self.registry_path)
            .context("Failed to rename registry file")?;

        Ok(())
    }
}

/// Cross-platform process termination
fn terminate_process(pid: u32, force: bool) -> Result<()> {
    #[cfg(unix)]
    {
        let signal = if force {
            Signal::SIGKILL
        } else {
            Signal::SIGTERM
        };
        
        let pid = Pid::from_raw(pid as i32);
        
        // Try to kill process group first (negative PID)
        let group_result = kill(Pid::from_raw(-pid.as_raw()), signal);
        
        if group_result.is_err() {
            // Fallback to single process
            kill(pid, signal)
                .context("Failed to send signal to process")?;
        }
        
        Ok(())
    }

    #[cfg(windows)]
    {
        use std::process::Command;
        
        let force_flag = if force { "/F" } else { "/T" };
        
        let output = Command::new("taskkill")
            .args(&[force_flag, "/PID", &pid.to_string()])
            .output()
            .context("Failed to execute taskkill")?;

        if !output.status.success() {
            let code = output.status.code().unwrap_or(-1);
            // Code 128 means process not found (already terminated)
            if code != 128 {
                return Err(anyhow::anyhow!(
                    "taskkill failed with code {}: {}",
                    code,
                    String::from_utf8_lossy(&output.stderr)
                ));
            }
        }

        Ok(())
    }
}

/// Check if error indicates process not found
fn is_process_not_found_error(error: &anyhow::Error) -> bool {
    let error_str = error.to_string().to_lowercase();
    error_str.contains("no such process") || error_str.contains("esrch")
}

// ============================================================================
// Tests
// ============================================================================

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    #[tokio::test]
    async fn test_registry_initialization() {
        let temp_dir = TempDir::new().unwrap();
        let registry_path = temp_dir.path().join("session.json");
        
        let mut registry = ProcessRegistry::new("test-session".to_string(), registry_path);
        registry.initialize().await.unwrap();

        assert_eq!(registry.registry.session_id, "test-session");
        assert_eq!(registry.registry.status, SessionStatus::Active);
    }

    #[tokio::test]
    async fn test_register_process() {
        let temp_dir = TempDir::new().unwrap();
        let registry_path = temp_dir.path().join("session.json");
        
        let mut registry = ProcessRegistry::new("test-session".to_string(), registry_path);
        registry.initialize().await.unwrap();

        registry
            .register(12345, "cursor", "echo", vec!["hello".to_string()])
            .await
            .unwrap();

        assert_eq!(registry.registry.processes.len(), 1);
        let process = &registry.registry.processes[0];
        assert_eq!(process.pid, 12345);
        assert_eq!(process.platform, "cursor");
        assert_eq!(process.status, ProcessStatus::Running);
    }

    #[tokio::test]
    async fn test_unregister_process() {
        let temp_dir = TempDir::new().unwrap();
        let registry_path = temp_dir.path().join("session.json");
        
        let mut registry = ProcessRegistry::new("test-session".to_string(), registry_path);
        registry.initialize().await.unwrap();

        registry
            .register(12345, "cursor", "echo", vec![])
            .await
            .unwrap();

        registry.unregister(12345, Some(0)).await.unwrap();

        let process = &registry.registry.processes[0];
        assert_eq!(process.status, ProcessStatus::Exited);
        assert_eq!(process.exit_code, Some(0));
    }

    #[tokio::test]
    async fn test_list_active() {
        let temp_dir = TempDir::new().unwrap();
        let registry_path = temp_dir.path().join("session.json");
        
        let mut registry = ProcessRegistry::new("test-session".to_string(), registry_path);
        registry.initialize().await.unwrap();

        // Register multiple processes
        registry.register(100, "cursor", "cmd1", vec![]).await.unwrap();
        registry.register(200, "claude", "cmd2", vec![]).await.unwrap();
        registry.unregister(100, Some(0)).await.unwrap();

        let active = registry.list_active();
        assert_eq!(active.len(), 1);
        assert_eq!(active[0].pid, 200);
    }

    #[tokio::test]
    async fn test_list_by_platform() {
        let temp_dir = TempDir::new().unwrap();
        let registry_path = temp_dir.path().join("session.json");
        
        let mut registry = ProcessRegistry::new("test-session".to_string(), registry_path);
        registry.initialize().await.unwrap();

        registry.register(100, "cursor", "cmd1", vec![]).await.unwrap();
        registry.register(200, "claude", "cmd2", vec![]).await.unwrap();
        registry.register(300, "cursor", "cmd3", vec![]).await.unwrap();

        let cursor_procs = registry.list_by_platform("cursor");
        assert_eq!(cursor_procs.len(), 2);
    }

    #[tokio::test]
    async fn test_mark_crashed() {
        let temp_dir = TempDir::new().unwrap();
        let registry_path = temp_dir.path().join("session.json");
        
        let mut registry = ProcessRegistry::new("test-session".to_string(), registry_path);
        registry.initialize().await.unwrap();

        registry.mark_crashed().await.unwrap();
        assert_eq!(registry.registry.status, SessionStatus::Crashed);
    }

    #[tokio::test]
    async fn test_persistence() {
        let temp_dir = TempDir::new().unwrap();
        let registry_path = temp_dir.path().join("session.json");
        
        // Create and save
        {
            let mut registry = ProcessRegistry::new("test-session".to_string(), registry_path.clone());
            registry.initialize().await.unwrap();
            registry.register(12345, "cursor", "echo", vec![]).await.unwrap();
        }

        // Load and verify
        {
            let mut registry = ProcessRegistry::new("test-session".to_string(), registry_path);
            registry.initialize().await.unwrap();
            assert_eq!(registry.registry.processes.len(), 1);
            assert_eq!(registry.registry.processes[0].pid, 12345);
        }
    }
}
